module/VuFind/src/VuFind/Controller/Plugin/AbstractRequestBase.php

```diff
@@ -303,6 +316,76 @@ public function getDefaultRequiredDate($checkHolds, $catalog = null,
         return $this->getDateFromArray($dateArray);
     }
 
+    /**
+     * Check if the user-provided dates are valid.
+     *
+     * Returns validated dates and/or an array of validation errors if there are
+     * problems.
+     *
+     * @param string $startDate       User-specified start date
+     * @param string $requiredBy      User-specified required-by date
+     * @param array  $extraHoldFields Hold form fields enabled by
+     * configuration/driver
+     *
+     * @return array
+     */
+    public function validateDates($startDate, $requiredBy, $extraHoldFields)
+    {
+        $result = [
+            'startDateTS' => null,
+            'requiredByTS' => null,
+            'errors' => [],
+        ];
+        if (!in_array('startDate', $extraHoldFields)
+            && !in_array('requiredByDate', $extraHoldFields)
+        ) {
+            return $result;
+        }
+
+        $errors = [];
+        if (in_array('startDate', $extraHoldFields)) {
+            try {
+                $result['startDateTS'] = $startDate
+                    ? $this->dateConverter->convertFromDisplayDate('U', $startDate)
+                    : 0;
+                if ($result['startDateTS'] < time()) {
+                    $errors[] = 'hold_start_date_invalid';
+                }
+            } catch (DateException $e) {
+                $errors[] = 'hold_start_date_invalid';
+            }
+        }
+
+        if (in_array('requiredByDate', $extraHoldFields)) {
+            try {
+                if ($requiredBy) {
+                    $result['requiredByTS'] = $this->dateConverter
+                        ->convertFromDisplayDate('U', $requiredBy, true)
+                        ->setTime(23, 59, 59)
+                        ->getTimestamp();
+                } else {
+                    $result['requiredByTS'] = 0;
+                }
+                if ($result['requiredByTS'] < time()) {
+                    $errors[] = 'hold_required_by_date_invalid';
```

Comment:
And these error messages also would need to be passed in somehow -- which I think further suggests the idea of a more complex input array, like `['requiredBy' => ['value' => ...,, 'error_msg' => ..., 'options' => ...]`. Maybe that's over complicated, but I think it could pay off in the long run.